packages/@react-aria/interactions/src/useFocusVisible.ts

```diff
@@ -194,26 +194,30 @@ export function useInteractionModality(): Modality {
  * Manages focus visible state for the page, and subscribes individual components for updates.
  */
 export function useFocusVisible(props: FocusVisibleProps = {}): FocusVisibleResult {
-  setupGlobalFocusEvents();
-
   let {isTextInput, autoFocus} = props;
   let [isFocusVisibleState, setFocusVisible] = useState(autoFocus || isFocusVisible());
-  useEffect(() => {
-    let handler = (modality: Modality, e: HandlerEvent) => {
-      // If this is a text input component, don't update the focus visible style when
-      // typing except for when the Tab and Escape keys are pressed.
-      if (isTextInput && modality === 'keyboard' && e instanceof KeyboardEvent && !FOCUS_VISIBLE_INPUT_KEYS[e.key]) {
-        return;
-      }
-
-      setFocusVisible(isFocusVisible());
-    };
-
-    changeHandlers.add(handler);
-    return () => {
-      changeHandlers.delete(handler);
-    };
+  useFocusEmitter((isFocusVisible, modality, e) => {
+    // If this is a text input component, don't update the focus visible style when
+    // typing except for when the Tab and Escape keys are pressed.
+    if (isTextInput && modality === 'keyboard' && e instanceof KeyboardEvent && !FOCUS_VISIBLE_INPUT_KEYS[e.key]) {
+      return;
+    }
+
+    setFocusVisible(isFocusVisible);
   }, [isTextInput]);
 
   return {isFocusVisible: isFocusVisibleState};
 }
+
+/**
+ * Emits on on trigger change and reports if focus is visible (i.e., modality is not pointer).
+ */
+export function useFocusEmitter(fn, deps): void {
```

Comment:
didn't know how to name this new hook, so we can definitely change it to something better

Reply:
IMHO `useFocusVisibilityEmitter` feels better to me, more specific that way. Note: names are hard and I suck at them